src/app/zones/data/index.ts

```diff
@@ -12,25 +22,16 @@ export function getZoneDpServerAuthType(zone: ZoneOverview): string {
   }
   return ''
 }
+// TODO(jc): end
 
-export function getZoneControlPlaneStatus(zoneOverview: ZoneOverview): StatusKeyword | 'disabled' {
+export function getZoneControlPlaneStatus(zoneOverview: ZoneOverview): 'online' | 'offline' | 'disabled' {
+  // TODO(jc): should I check disabled for newly created zones for the purposes of this PR?
   if (zoneOverview.zone.enabled === false) {
     return 'disabled'
   }
-
-  const subscriptions = zoneOverview.zoneInsight?.subscriptions ?? []
-  if (subscriptions.length === 0) {
-    return 'offline'
-  }
-
-  const lastSubscription = subscriptions[subscriptions.length - 1]
-  return lastSubscription.connectTime?.length && !lastSubscription.disconnectTime ? 'online' : 'offline'
+  return getStatus(zoneOverview.zoneInsight?.subscriptions)
 }
-export function getZoneControlPlaneEnvironment(zoneOverview: ZoneOverview): string {
-  for (const subscription of zoneOverview.zoneInsight?.subscriptions ?? []) {
-    if (subscription.config) {
-      return JSON.parse(subscription.config).environment
-    }
-  }
-  return ''
+export function getStatus(subscriptions: {connectTime?: string, disconnectTime?: string}[] | undefined = []): 'online' | 'offline' {
+  const proxyOnline = subscriptions.length > 0 && [subscriptions[subscriptions.length - 1]].every((item) => item.connectTime?.length && !item.disconnectTime)
+  return proxyOnline ? 'online' : 'offline'
```

Comment:
**Issue**: Why is this implementation being changed?

I think this is significantly harder to parse and understand than the previous code:

```ts
  const subscriptions = zoneOverview.zoneInsight?.subscriptions ?? []
  if (subscriptions.length === 0) {
    return 'offline'
  }

  const lastSubscription = subscriptions[subscriptions.length - 1]
  return lastSubscription.connectTime?.length && !lastSubscription.disconnectTime ? 'online' : 'offline'
```

Especially the `[subscriptions[subscriptions.length - 1]].every((item) => item.connectTime?.length && !item.disconnectTime)` bit is very convoluted. Why would we create an array of definite length 1 to use `Array.prototype.every` on it?

Besides this, why are we separating the online/offline check from the disabled check? Seems to me that all three form a Zone’s status.

Reply:
There are quite a few things here so bear with me while I try separate them a little.

> 1. Issue: Why is this implementation being changed?
> Besides this, why are we separating the online/offline check from the disabled check? Seems to me that all three form a Zone’s status.

Previous to this, the code was using the `getStatusFromInsight` function in the `utilities/dataplane` file.

removal: https://github.com/kumahq/kuma-gui/pull/1634/files#diff-9cd2f8af449e6e0679c6c3ec12e5690e7fdc80a964ac738c7626cd5bc0dc6ac7L388

https://github.com/kumahq/kuma-gui/blob/6d6ac8a18925202d728e0300dab2ef5be55b1e23/src/utilities/dataplane.ts#L71-L74

This code is about zones not dataplanes, so copied the code and moved it to zones. When I did that I saw that there was another `getZoneControlPlaneStatus` function that already existed, but I wasn't totally sure what that is used for / why we were using a different dataplane related function in the first place. I also noticed that I could reuse the code that I'd copied from dataplanes (which was essentially a `subscription` util) so I decided to move/merge what I could that made sense, and then stop and ask questions on the rest, hence in this PR:

> Should I check disabled for newly created zones for the purposes of this PR? (this one should be done here if necessary)

This is not what it was doing previously:

https://github.com/kumahq/kuma-gui/commit/1afcaacf6ba11e2eb3195290a61961c77d97a60f

It sounds like from your response here that we should be checking the `disabled` attribute also which makes sense to me and also means I can just reuse.

Does that sound right to you? i.e. we should check disabled for newly created zones for the purposes of this PR?

Reply:
I think, yes. A disabled Zone is neither offline or online or at least for offline, the difference is philosophical.

Reply:
I changed the callsite in `source.ts` to use `getZoneControlPlaneStatus` instead of just `getStatus`. This means that creation also checks for the zone to be `enabled`. For the moment I left everything else as is, to be amended in a more refactor-ful PR.

I'd probably like to change the naming of `getZoneControlPlaneStatus` seeing as its in the `zones` module, move `getStatus` to a new `subscriptions` module, plus the amends mentioned in the comments and TODOs here.

Reply:
I think I can resolve this now, but as mentioned there is a little moving around and probably a bit of renaming and reuse to be done in an upcoming PR. but, let me know if this isn't resolved.

Reply:
The initial reason why I opened this thread is not addressed, yet.

The logic for the online/offline check was changed from this:

```ts
  if (subscriptions.length === 0) {
    return 'offline'
  }

  const lastSubscription = subscriptions[subscriptions.length - 1]
  return lastSubscription.connectTime?.length && !lastSubscription.disconnectTime ? 'online' : 'offline'
```

to this:

```ts
  const proxyOnline = subscriptions.length > 0 && [subscriptions[subscriptions.length - 1]].every((item) => item.connectTime?.length && !item.disconnectTime)
  return proxyOnline ? 'online' : 'offline'
```

And I think that makes it unnecessarily hard to understand what the logic actually is. Especially the `[subscriptions[subscriptions.length - 1]].every((item) => item.connectTime?.length && !item.disconnectTime)` is really weird to look at.

Can we please move this back?

Reply:
Oh sorry, well I pretty much copied this code from:

https://github.com/kumahq/kuma-gui/blob/master/src/utilities/dataplane.ts#L72

```javascript
const proxyOnline = (insight.subscriptions ?? []).some((subscription) => subscription.connectTime?.length && !subscription.disconnectTime)
```

I've kept it as similar as possible, I've not really not changed it and i would rather keep it as similar as possible, until I get to do the follow up PR here that I keep talking about.

As I keep saying here, I'll be doing a more general cleanup of this zone data layer file once I get this PR merged. That PR will involve me looking at all the code there and reducing the places where it does the same thing but slightly differently and removing all the places that use slightly different functions.

Honestly considering I've been saying I'll doing all this in a follow I don't think this is something to be concerned about, especially considering I'm pretty much using the same code as it was before this PR.

Finally, I think I prefer the code as a one-liner, not that that matters too much at all.

Reply:
If you’re changing this back in a follow up, I’m fine with that so feel free to resolve this thread.

One liners are great if they easily convey what’s happening. Neither `[subscriptions[subscriptions.length - 1]].every((item) => item.connectTime?.length && !item.disconnectTime)` nor `(insight.subscriptions ?? []).some((subscription) => subscription.connectTime?.length && !subscription.disconnectTime)` qualify for this, I’m afraid.